### ISSUE TITLE
LambdaType should be renamed ArrowType

### ISSUE THREAD

**Issue by kyouko-taiga** [hylo-lang/hylo#385]
`LambdaType` is used to type all "arrow-like" objects in the language, not only lambdas. The only exception are method bundles, but those already have a dedicated `MethodType`.

**Comment by dabrahams** [comment]
What are some other arrow-like objects?

**Comment by kyouko-taiga** [comment]
```swift
type A {
  init() {}
  fun ham() -> Int { 2 }
}

public fun main() {
  fun foo() -> Int { 0 }
  let bar = fun() -> Int { 1 }
  let a = A()
  ...
}
```

`foo`, `bar`, `A.init`, `A.ham`, and `a.ham` are arrow-like objects, but they're not all the exact same kind of arrows. That has subtle consequences on the semantics.

Note that only `bar` is a lambda.

**Comment by dabrahams** [comment]
Is there any role for the word “callable” in addressing this issue?

**Comment by kyouko-taiga** [comment]
Could be `CallableType` instead of `ArrowType`. The latter is shorter and very a quite common name in type theory though.

**Comment by dabrahams** [comment]
Not just `Callable`?

**Comment by kyouko-taiga** [comment]
I'd like to keep the suffix, at least for now. I still think it's helping distinguishing between important kinds, as we have entities like `TraitType` and `TraitDecl`.

Given the way we tend to name stuff in Swift, I would also reserve `Callable` for a protocol describing a family of arrow types, should we have one.

**Comment by peter-evans** [comment]
I could give this rename a shot if it's helpful. Just want to check this issue is still relevant.

**Comment by kyouko-taiga** [comment]
I think it is.

Note that now we have `CallableType` which is implemented by the arrow types _and_ `SubscriptType`, filling the role of the trait I hypothesized we'd need:

> I would also reserve `Callable` for a protocol describing a family of arrow types, should we have one.

So you'll have to rename that one too.

**Comment by peter-evans** [comment]
Just to clarify, what would you like to rename `CallableType` to?

`LambdaType` -> `ArrowType`
`CallableType` -> ?

**Comment by kyouko-taiga** [comment]
Actually if we rename `LambdaType` to `ArrowType` then there's nothing to change for `CallableType`. I'd accept that change.

@dabrahams had expressed reluctance because (IIUC) "arrow" isn't as established as "callable". But I think this ship has sailed because the distinction is already significant in the compiler in other places:

https://github.com/hylo-lang/hylo/blob/9ade3acff6c34f8041c8e6306105a0797cf2814a/Sources/Core/Types/CallableType.swift#L10-L11

https://github.com/hylo-lang/hylo/blob/9ade3acff6c34f8041c8e6306105a0797cf2814a/Sources/FrontEnd/TypeChecking/Constraints/CallConstraint.swift#L35-L36

So to be clear, I'm suggesting that you only rename `LambdaType` to `ArrowType`.

**Comment by dabrahams** [comment]
"Arrow" remains very esoteric and *if* another term like “lambda” could work, I'd prefer it.  Just looking around for definitions of arrow in this domain it's very hard to find references that don't lead to something like [this](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions#).  Just trying to refresh my memory about what an Arrow is I am still not quite sure.  [One definition](http://www.euclideanspace.com/maths/discrete/category/principles/arrow/index.htm) says it's a "structure preserving" map which makes me think in general functions don't fit the category since they can be many-to-one, and that's confusing.  Maybe I'm too old to use _the google_ but I'm still looking for a simple definition that I can stand on.

If this is the only correct term, of course, so be it.  But at a minimum docs for names that include the term should contain links to a comprehensible definition.

**Comment by dabrahams** [comment]
_I_ can understand [this](https://www.haskell.org/arrows/), but only because I understand Haskell notation… and what the "number of axioms" they mention are is still a mystery.  IIUC I would sum the arrow idea up as "functions within a domain," where the domain (best word I can think of) is not the idea related to _codomain_ but is the type class `a` itself.  I don't know what the multiple domains might be in practice but regardless don't we have just one of those, and isn't the domain (`a`) the thing that _is_ the arrow, as opposed to particular instances of `a`?

**Comment by kyouko-taiga** [comment]
I understand the confusion and sympathize with it. Naming is hard and I'm not particularly good at it.

We do have a constellation of concepts that needs different names, though:

1. Things that are callable, which include subscript bundles, subscripts, method bundles, functions, initializers, and lambdas
2. Things that create projections, which include subscript bundles and subscripts
3. Things that return values, which include functions, initializers, and lambdas

`Callable` was a strong candidate for (2) but:
- entities in (3) are also "called", so ultimately I don't think it's right to reserve the term for (2)
- we'd still need a term for (1)

`Application` could be a candidate for (1), being very well established in lambda calculus, but I think the term is too overloaded in the context of a compiler implementation. I also think the term is much closer to lambdas than it is to subscripts, so I don't think it's the right fit.

I like "arrow" because it's tied to the way to express these type terms. An arrow is just a callable that is declared ... with an arrow. It's also short and concise; `if someType.isArrow { ... }` makes a lot of sense to me.

Note that whatever we choose, these terms are very internal to Hylo's frontend and only significant to the few souls courageous enough to venture in the type checker's implementation. So to some degree I think it's fine to just come up with our own vocabulary if there isn't strong precedent to guide us. We still have to write a lexicon, and these terms would be part of it.

**Comment by dabrahams** [comment]
> I like "arrow" because it's tied to the way to express the types in (2). An arrow is just a callable that is declared ... with an arrow. 

Oh, you mean syntactically.  Sure, but it _isn't_ just that syntactic arrow that is only _part_ of its declaration, and I'm still very unclear on all the points about how (or even whether) the category-theoretic idea maps on to this thing.

> `if someType.isArrow { ... }` makes a lot of sense to me.

Well sure, but the problem I'm trying to grapple with is whether it's going to make sense to anyone else.  What's wrong with `if someType.returnsAValue { ... }` which is literally how you described this category in plain English?  If you really insist on something that's both short _and_ concise (😉) I offer `someType.returns { ... }` but personally I don't think the concision is worth the loss of straightforward comprehensibility.

And if you need a noun… from my point of view, “functions, initializers, and lambdas” (don't forget methods) are all just functions with various declaration syntaxes, and “function” would make a much better term for these things.  IIRC Swift named even top-level functions "methods"—since they could easily be viewed “static methods of the module”—and doing that would definitively free up the term “function” in the compiler.

> Note that whatever we choose, these terms are very internal to Hylo's frontend and only significant to the few souls courageous enough to venture in the type checker's implementation

Really?  It seems to me we want to use the same terms in our specification and instructional materials that we use in the compiler implementation to classify the language's major entities.

**Comment by peter-evans** [comment]
Just some observations as a novice looking at this codebase. If I'm muddying the conversation feel free to ignore. 😄 

> "Arrow" remains very esoteric and if another term like “lambda” could work, I'd prefer it.

It's clear to me that it would be best to change the name from "lambda" to *something*, because it's been difficult in some cases to figure out if the code is referring to arrow-like expressions, or to *actual* lambda expressions.

> Sure, but it isn't just that syntactic arrow that is only part of its declaration

I see this point. If I was looking at codebase for the first time it might not be clear that "arrow" doesn't just refer to `->` but encompasses the whole declaration of the arrow-like expression. Perhaps `ArrowlikeType` and `isArrowlike()` would be preferable to just `ArrowType`.

> 1. Things that are callable, which include subscript bundles, subscripts, method bundles, functions, initializers, and lambdas
> 2. Things that return values, which include functions, initializers, and lambdas
> 3. Things that create projections, which include subscript bundles and subscripts

Here's my very naive suggestion that would be clear what each type does:
1. `Callable`
2. `CallableWithReturn`
3. `CallableWithProjection`

**Comment by kyouko-taiga** [comment]
It's my bad having described (2) and (3) that way, but we can't use `returns`, `returnsAValue`, or `callableWithReturn` to identify arrow types. `sink`-subscripts return a value and they are not arrow-like. Yet they must satisfy the same predicate as `let`-subscripts.

> IIRC Swift named even top-level functions "methods"

I didn't forget methods. In fact I made the same simplification as the Swift crowd, except that I decided to refer to methods as functions, since they are just functions with an environment.

I once considered using `isFunction` and `isSubscript` but the distinction between (2) and (3) is still significant in the IR, where all subscripts have been lowered as functions!

**Comment by dabrahams** [comment]
You keep calling them arrow types as though that's what they _are_.  If they really are that, you should be able to clear up all my questions about the category-theoretic mapping.  But you haven't done that, so I feel both like there's a fâit accompli here, and like I'm not being heard.

> I didn't forget methods. In fact I made the same simplification as the Swift crowd, except that I decided to refer to methods as simply functions, since they are just functions with an environment. 

Everything has an environment.  It just might be void.

> sink-subscripts return a value and they are not arrow-like.

I'm increasingly frustrated with the way the word “arrow” is being used here without explanation or clarification, when I've made it quite apparent that its meaning in this context is unclear.  In what _way_ are sink subscripts less “arrow-like” than a function taking a sink parameter?

> So AFAIC, a method is just another name for a "non-static member function". Method bundles are a different concept, though. I once considered using isFunction and isSubscript but the distinction between (2) and (3) is still significant in the IR, where all subscripts have been lowered as functions!

You're missing my point.  There's a way to shuffle the names around so that "Function" becomes the name of category (2) and you still have subscripts (or projections).  You just have to be willing to view the things you're currently thinking of as "plain functions" as static methods of the module.  IMO that would be preferable to using a term that you haven't even been able to help _me_ understand, to say nothing of the rest of the world.

**Comment by kyouko-taiga** [comment]
The frustration is shared. The name has nothing to do with category theory. I explained where it comes from. The type term has an arrow, that's it. I don't know how to explain that better.

The fact that a tiny fraction of Haskell enthusiast use "arrow type" to express monad-like abstractions is coincidental but name clashes do happen.

> Everything has an environment. It just might be void.

Of course I know that everything has an environment. It should be equally obvious that a lambda with an "empty" environment is equivalent to a lambda with a "void" environment, which is an empty type.

>  In what way are sink subscripts less “arrow-like” than a function taking a sink parameter?

Subscripts are not written with an arrow in their type. The compiler cares about that because it must generate different code for a `sink`-subscript than it does for a function.

> You just have to be willing to view the things you're currently thinking of as "plain functions" as static methods of the module. IMO that would be preferable to using a term that you haven't even been able to help me understand, to say nothing of the rest of the world.

It is frustrating to have to change _my_ vocabulary and way of thinking about entities, which are used consistently in the many lines of the implementation that I mostly wrote and maintain, because other people prefer other terms. The _faits accomplis_ come from the fact that I keep having to make choices for representing pretty much everything in the compiler, and I can't pause every time to setup a vote.

I'm happy to explain every possible gory detail of the compiler to anyone who'd like me to, really. But I find it frustrating to be pressed for _justifying_ why something has been written the way out of the blue.

**Comment by dabrahams** [comment]
> The name has nothing to do with category theory

I can't understand why you didn't ever say that earlier.  I kept going down that path and you never stopped me… 😢 

> I explained where it comes from. The type term has an arrow, that's it. I don't know how to explain that better.

You didn't say it had to do with type terms; you said the arrow referred to what the entity was “declared with,” and allowed me to assert that it was about declaration syntax without correcting me.  I see now that in fact you're not saying it has anything to do with the declaration syntax of the entity; it's about the syntax one would use to express the type (which is never declared, because it's not nominal).  I realize you dropped the hint "it's tied to the way to express the types" but that's pretty vague and the correct interpretation of that reference is ruled out by what you said about declaration immediately following that.

> The fact that a tiny fraction of Haskell enthusiast use "arrow type" to express monad-like abstractions is coincidental but name clashes do happen.

This potential confusion is very easily solved.  Just stop calling these things arrows and say that they're types that _have_ arrows.  If you really must.

> Subscripts are not written with an arrow in their type. 

Subscripts don't have a type as far as I know.  At least I've never seen a subscript type in any example we've discussed.  I think you're talking about declaration syntax again.  But then intializers don't have an arrow in their declaration syntax either.

> It is frustrating to have to change my vocabulary and way of thinking about entities, which are used consistently in the many lines of the implementation that I mostly wrote and maintain, because other people prefer other terms.

It's not about my preference; it's about understanding.  If I can't understand the term and it leads me to look for connections _that aren't even supposed to be there_, the project is harder than necessary for others to use, participate in, and contribute to.  I assert _again_ that we will want to use the same terms in our specification and instructional materials that we use in the compiler implementation to classify the language's major entities.

**Comment by sean-parent** [comment]
> We do have a constellation of concepts that needs different names, though:
> 
> 1. Things that are callable, which include subscript bundles, subscripts, method bundles, functions, initializers, and lambdas
> 2. Things that return values, which include functions, initializers, and lambdas
> 3. Things that create projections, which include subscript bundles and subscripts

Questions:

What are the current names of these concepts, and which one is being considered to be remapped to "arrow"?

For the items that are callable and don't create a projection or return a value, what do they result in?

**Comment by kyouko-taiga** [comment]
> What are the current names of these concepts, and which one is being considered to be remapped to "arrow"?

Functions are declared by `FunctionDecl` (see [here](https://github.com/hylo-lang/hylo/blob/5b773b79f9ea009c7480625c3b8021cb74817dc9/Sources/Core/AST/Decl/FunctionDecl.swift)); that includes the declarations of anonymous functions underlying lambdas;
Initializers are declared by `InitializerDecl`  (see [here](https://github.com/hylo-lang/hylo/blob/5b773b79f9ea009c7480625c3b8021cb74817dc9/Sources/Core/AST/Decl/InitializerDecl.swift));
Method bundles and method variants (i.e., the individual implementations of a bundle) are declared by `MethodDecl`  (see [here](https://github.com/hylo-lang/hylo/blob/5b773b79f9ea009c7480625c3b8021cb74817dc9/Sources/Core/AST/Decl/MethodDecl.swift)) and `MethodImpl`  (see [here](https://github.com/hylo-lang/hylo/blob/5b773b79f9ea009c7480625c3b8021cb74817dc9/Sources/Core/AST/Decl/MethodImpl.swift)), respectively;
Subscript bundles and subscript variants are declared by `SubscriptDecl` (see [here](https://github.com/hylo-lang/hylo/blob/5b773b79f9ea009c7480625c3b8021cb74817dc9/Sources/Core/AST/Decl/SubscriptDecl.swift)) and `SubscriptImpl`  (see [here](https://github.com/hylo-lang/hylo/blob/5b773b79f9ea009c7480625c3b8021cb74817dc9/Sources/Core/AST/Decl/SubscriptImpl.swift)), respectively.

All of these declarations introduce a _callable_ entity (see [here](https://github.com/hylo-lang/hylo/blob/5b773b79f9ea009c7480625c3b8021cb74817dc9/Sources/Core/AST/Decl/Decl.swift#L8)).

Functions, initializers, and method variants are given a `LambdaType` (see [here](https://github.com/hylo-lang/hylo/blob/5b773b79f9ea009c7480625c3b8021cb74817dc9/Sources/Core/Types/LambdaType.swift)).
Method bundles are given a `MethodType` (see [here](https://github.com/hylo-lang/hylo/blob/5b773b79f9ea009c7480625c3b8021cb74817dc9/Sources/Core/Types/MethodType.swift)).
Subscript bundles are given a `SubscriptType` (see [here](https://github.com/hylo-lang/hylo/blob/5b773b79f9ea009c7480625c3b8021cb74817dc9/Sources/Core/Types/SubscriptType.swift)).
Subscript variants are given a `SubscriptImplType` (see [here](https://github.com/hylo-lang/hylo/blob/5b773b79f9ea009c7480625c3b8021cb74817dc9/Sources/Core/Types/SubscriptImplType.swift)).

All of these types are `CallableType`s (see [here](https://github.com/hylo-lang/hylo/blob/5b773b79f9ea009c7480625c3b8021cb74817dc9/Sources/Core/Types/CallableType.swift)).

A callable type is considered an _arrow_ if it's the type of a function, initializer, or method variant (see [here](https://github.com/hylo-lang/hylo/blob/5b773b79f9ea009c7480625c3b8021cb74817dc9/Sources/Core/Types/CallableType.swift#L11)). Types for which this property holds are `LambdaType` and `MethodType`.

The change proposed by the issue is to rename `LambdaType` to `ArrowType`.

> For the items that are callable and don't create a projection or return a value, what do they result in?

There are no such items. Functions, initializers, and methods always _return_ a value as far as the type system is concerned. That value may be an instance of `Void` or `Never`. Subscripts _project_ a value along with a capability. That capability may be `sink`, making the projection notionally equivalent to a returned value.

While one may think of a sink subscript as being equivalent to a function, the type checker reasons differently because it has to consider the effect on the arguments of the subscript. I'm happy to get into more details but I hope it would create distractions at this point.
